<a href="https://colab.research.google.com/github/Assaoka/Decolar--Introducao_a_Ciencia_de_Dados/blob/main/%5B%20%206%20%5D%20Sum%C3%A1rios%20e%20Transforma%C3%A7%C3%B5es/Sum%C3%A1rios%20e%20Transforma%C3%A7%C3%B5es.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Aula 6: A Análise do Mestre de Dados: Sumários, Filtros Avançados e Transformações

Na nossa última jornada (Aula 5), você deu um passo gigante: aprendeu a usar o Pandas para capturar Pokémon (ou seja, carregar dados de um arquivo) e a dar uma primeira espiada neles com comandos como `.head()` e `.tail()`.

Hoje, vamos aprofundar nossas habilidades de análise. Você vai aprender a fazer perguntas muito mais inteligentes e específicas para a nossa Pokédex e até a criar novas informações sobre os Pokémon que não existiam no arquivo original. Vamos lá!

### **1. Revisão Rápida: Relembrando nossos Poderes**

Antes de aprender novos movimentos, todo bom treinador revisa os que já conhece. Vamos relembrar o que fizemos na última aula:

1. **Importamos o Pandas:** O nosso kit de ferramentas. `import pandas as pd`
    
2. **Carregamos a Pokédex:** Usamos o `pd.read_csv('pokemon.csv')` para ler o arquivo.
    
3. **Inspecionamos os Dados:** Usamos `.head()` para ver o começo, `.tail()` para ver o fim e `.info()` para ver a estrutura da nossa tabela.
    
4. **Fizemos Filtros Simples:** Aprendemos a encontrar Pokémon com uma característica específica, como `df[df['Primary Typing'] == 'fire']`.
    

Vamos começar carregando nossa Pokédex novamente. Execute o código abaixo:

In [1]:
# Importando nossa biblioteca principal e dando o apelido de 'pd'
import pandas as pd

In [2]:
# Endereço do nosso arquivo de dados
endereco = 'https://raw.githubusercontent.com/Assaoka/Decolar--Introducao_a_Ciencia_de_Dados/refs/heads/main/pokemon.csv'

# Lendo o arquivo CSV e guardando na variável 'df' (nosso DataFrame)
df = pd.read_csv(endereco)

In [3]:
# Mostrando os 5 primeiros Pokémon para garantir que tudo deu certo
display(df.head())

,Name,National Dex #,Primary Typing,Secondary Typing,Secondary Typing Flag,Generation,Legendary Status,Form,Alt Form Flag,Evolution Stage,...,Weight (hg),Height (in),Weight (lbs),Base Stat Total,Health,Attack,Defense,Special Attack,Special Defense,Speed
0,bulbasaur,1,grass,poison,True,generation-i,False,Base,False,1,...,69,28,15,318,45,49,49,65,65,45
1,ivysaur,2,grass,poison,True,generation-i,False,Base,False,2,...,130,39,29,405,60,62,63,80,80,60
2,venusaur,3,grass,poison,True,generation-i,False,Base,False,3,...,1000,79,220,525,80,82,83,100,100,80
3,venusaur-mega,3,grass,poison,True,generation-i,True,Mega,True,3,...,1555,94,343,625,80,100,123,122,120,80
4,charmander,4,fire,NaN,False,generation-i,False,Base,False,1,...,85,24,19,309,39,52,43,60,50,65


### **2. A Ficha Técnica da Pokédex (`.describe()`)**

Imagine que você pudesse pedir ao Professor Carvalho um resumo completo sobre os poderes de todos os Pokémon de uma só vez. Em vez de olhar um por um, ele te entregaria uma ficha com a **média** de força, o Pokémon **mais forte**, o **mais fraco** e muito mais.

É exatamente isso que o comando `.describe()` faz para as colunas que têm números.

In [4]:
# Usando .describe() para analisar os stats de batalha
# Selecionamos apenas as colunas que queremos analisar
df[['Attack', 'Defense', 'Speed', 'Base Stat Total']].describe()

,Attack,Defense,Speed,Base Stat Total
count,1184.000000,1184.000000,1184.000000,1184.000000
mean,80.989020,74.817568,69.728885,441.633446
std,31.955337,30.324208,30.110391,119.296702
min,5.000000,5.000000,5.000000,175.000000
25%,57.000000,52.000000,45.000000,330.000000
50%,80.000000,70.000000,67.500000,464.500000
75%,100.000000,91.000000,91.000000,520.000000
max,190.000000,230.000000,200.000000,780.000000


**O que significa cada linha desse resumo?**

- `count`: A contagem de quantos Pokémon têm um valor para aquele stat. Se o número for igual para todos, significa que não temos dados faltando!
    
- `mean` (**Média**): Se a gente somasse o "Attack" de TODOS os Pokémon e dividisse pelo número de Pokémon, chegaríamos nesse valor. É o valor "típico" de ataque.
    
- `std` (**Desvio Padrão**): Esse nome é um pouco assustador, mas a ideia é simples!
    
    - **Se o número for baixo:** Significa que a maioria dos Pokémon tem um "Attack" bem parecido com a média. Não há muita variação.
        
    - **Se o número for alto:** Significa que os valores de "Attack" são muito espalhados. Existem Pokémon super fracos e outros super fortes, bem longe da média.
        
- `min`: O menor valor encontrado. É o "Attack" do Pokémon mais fraco.
    
- `25%` (Primeiro Quartil): Se a gente colocar todos os Pokémon em uma fila, do mais fraco ao mais forte, este é o valor do Pokémon que está a um quarto do caminho. Ele é mais forte que 25% de todos os outros.
    
- `50%` (Mediana): Este é o Pokémon que está EXATAMENTE no meio da fila. Metade dos Pokémon é mais fraca que ele, e a outra metade é mais forte.
    
- `75%` (Terceiro Quartil): É o valor do Pokémon que está a três quartos do caminho na fila. Ele é mais forte que 75% de todos os outros.
    
- `max`: O maior valor encontrado. É o "Attack" do Pokémon mais forte de todos!
    

### **3. Contando Tipos de Pokémon (`.value_counts()`)**

Você já se perguntou qual o tipo de Pokémon mais comum? Água? Grama? Normal? Contar isso na mão seria impossível!

Para nossa sorte, temos o poder `.value_counts()`, que conta quantas vezes cada valor aparece em uma coluna. É perfeito para colunas de texto, como 'Primary Typing'.

In [5]:
# Contando quantos Pokémon existem para cada tipo primário
df['Primary Typing'].value_counts()

,count
Primary Typing,
water,145
normal,128
grass,113
bug,89
fire,77
psychic,77
rock,71
electric,71
dark,57


Agora sabemos que o tipo "Water" (Água) é o mais comum de todos!

In [6]:
# Também podemos ver em porcentagem, para ter uma ideia da proporção
# O 'normalize=True' calcula a proporção (um número de 0 a 1)
# e nós multiplicamos por 100 para ver em porcentagem.
df['Primary Typing'].value_counts(normalize=True) * 100

,proportion
Primary Typing,
water,12.246622
normal,10.810811
grass,9.543919
bug,7.516892
fire,6.503378
psychic,6.503378
rock,5.996622
electric,5.996622
dark,4.814189


### **4. Filtros Avançados: Fazendo Perguntas de Mestre (`&`, `|`)**

Na última aula, você fez perguntas simples, como "Quais Pokémon são do tipo Fogo?".

Um Mestre de Dados faz perguntas mais complexas! Por exemplo:

- "Quais Pokémon são do tipo 'Grama' **E** também do tipo 'Veneno'?"
    
- "Quais Pokémon são Lendários **OU** têm um 'Base Stat Total' maior que 600?"
    

Para isso, usamos dois operadores mágicos:

- `&` (significa **E**): As duas condições precisam ser verdadeiras.
    
- `|` (significa **OU**): Pelo menos uma das condições precisa ser verdadeira.
    

**IMPORTANTE:** Cada condição que você cria precisa estar dentro de parênteses `()`!

In [ ]:
# Exemplo com E (&): Pokémon de Fogo E Voadores
condicao_fogo = (df['Primary Typing'] == 'fire')
condicao_voador = (df['Secondary Typing'] == 'flying')

df[condicao_fogo & condicao_voador]

,Name,National Dex #,Primary Typing,Secondary Typing,Secondary Typing Flag,Generation,Legendary Status,Form,Alt Form Flag,Evolution Stage,...,Weight (hg),Height (in),Weight (lbs),Base Stat Total,Health,Attack,Defense,Special Attack,Special Defense,Speed
6,charizard,6,fire,flying,True,generation-i,False,Base,False,3,...,905,67,200,534,78,84,78,109,85,100
8,charizard-mega-y,6,fire,flying,True,generation-i,True,Mega Y,True,3,...,1005,67,222,634,78,104,78,159,115,100
197,moltres,146,fire,flying,True,generation-i,True,Base,False,1,...,600,79,132,580,90,100,90,125,85,90
316,ho-oh,250,fire,flying,True,generation-ii,True,Base,False,1,...,1990,150,439,680,106,130,90,110,154,90
784,fletchinder,662,fire,flying,True,generation-vi,False,Base,False,2,...,160,28,35,382,62,73,55,56,52,84
785,talonflame,663,fire,flying,True,generation-vi,False,Base,False,3,...,245,47,54,499,78,81,71,74,69,126
873,oricorio-baile,741,fire,flying,True,generation-vii,False,Base,False,1,...,34,24,7,476,75,70,70,98,70,93


In [ ]:
# Versão resumida:
# Exemplo com E (&): Pokémon de Fogo E Voadores
df[(df['Primary Typing'] == 'fire') & (df['Secondary Typing'] == 'flying')]

,Name,National Dex #,Primary Typing,Secondary Typing,Secondary Typing Flag,Generation,Legendary Status,Form,Alt Form Flag,Evolution Stage,...,Weight (hg),Height (in),Weight (lbs),Base Stat Total,Health,Attack,Defense,Special Attack,Special Defense,Speed
6,charizard,6,fire,flying,True,generation-i,False,Base,False,3,...,905,67,200,534,78,84,78,109,85,100
8,charizard-mega-y,6,fire,flying,True,generation-i,True,Mega Y,True,3,...,1005,67,222,634,78,104,78,159,115,100
197,moltres,146,fire,flying,True,generation-i,True,Base,False,1,...,600,79,132,580,90,100,90,125,85,90
316,ho-oh,250,fire,flying,True,generation-ii,True,Base,False,1,...,1990,150,439,680,106,130,90,110,154,90
784,fletchinder,662,fire,flying,True,generation-vi,False,Base,False,2,...,160,28,35,382,62,73,55,56,52,84
785,talonflame,663,fire,flying,True,generation-vi,False,Base,False,3,...,245,47,54,499,78,81,71,74,69,126
873,oricorio-baile,741,fire,flying,True,generation-vii,False,Base,False,1,...,34,24,7,476,75,70,70,98,70,93


In [ ]:
# Exemplo com OU (|): Pokémon com ataque MUITO alto OU defesa MUITO alta
df[(df['Attack'] > 150) | (df['Defense'] > 150)]

,Name,National Dex #,Primary Typing,Secondary Typing,Secondary Typing Flag,Generation,Legendary Status,Form,Alt Form Flag,Evolution Stage,...,Weight (hg),Height (in),Weight (lbs),Base Stat Total,Health,Attack,Defense,Special Attack,Special Defense,Speed
109,slowbro-mega,80,water,psychic,True,generation-i,True,Mega,True,2,...,1200,79,265,590,95,75,180,130,80,30
124,cloyster,91,water,ice,True,generation-i,False,Base,False,2,...,1325,59,292,525,50,95,180,85,45,70
129,onix,95,rock,ground,True,generation-i,False,Base,False,1,...,2100,346,463,385,35,45,160,30,45,70
170,pinsir-mega,127,bug,flying,True,generation-i,True,Mega,True,1,...,590,67,130,600,65,155,120,65,90,105
177,gyarados-mega,130,water,dark,True,generation-i,True,Mega,True,2,...,3050,256,672,640,95,155,109,70,130,81
203,mewtwo-mega-x,150,psychic,fighting,True,generation-i,True,Mega X,True,1,...,1270,91,280,780,106,190,100,154,100,130
266,steelix,208,steel,ground,True,generation-ii,False,Base,False,2,...,4000,362,882,510,75,85,200,55,65,30
267,steelix-mega,208,steel,ground,True,generation-ii,True,Mega,True,2,...,7400,413,1631,610,75,125,230,55,95,30
274,shuckle,213,bug,rock,True,generation-ii,False,Base,False,1,...,205,24,45,505,20,10,230,10,230,5
276,heracross-mega,214,bug,fighting,True,generation-ii,True,Mega,True,1,...,625,67,138,600,80,185,115,40,105,75


### **5. Transformando Dados: Criando Novas Informações (`.map` e `.apply`)**

E se a gente quisesse criar uma informação nova que não existe na Pokédex? Por exemplo, traduzir a coluna `Legendary Status` (que está como `True` ou `False`) para textos mais amigáveis como "Sim" e "Não"?

Para isso, temos dois poderes incríveis: `.map()` e `.apply()`.

#### **5.1. Usando `.map()` para "Traduzir" Valores**

O `.map()` é perfeito para quando você quer substituir valores de uma coluna por outros, usando um **dicionário** como se fosse um tradutor.

In [ ]:
# 1. Criamos nosso "dicionário tradutor"
mapa_tradutor = {
    True: 'Sim',
    False: 'Não'
}

# 2. Usamos .map() para criar uma nova coluna "Lendario" com os valores traduzidos
df['Lendario'] = df['Legendary Status'].map(mapa_tradutor)
display(df[['Name', 'Legendary Status', 'Lendario']].sample(20))

,Name,Legendary Status,Lendario
26,raticate,False,Não
1170,ogerpon,True,Sim
969,corviknight,False,Não
1003,hattrem,False,Não
695,reuniclus,False,Não
766,kyurem-white,True,Sim
1026,dracozolt,False,Não
1175,hydrapple,False,Não
377,mawile-mega,True,Sim
219,noctowl,False,Não


#### **5.2. Usando `.apply()` para Aplicar Regras Complexas**

O `.apply()` é ainda mais poderoso. Ele permite aplicar uma **função** (um conjunto de regras que a gente cria) para cada valor de uma coluna.

Vamos criar uma regra para classificar os Pokémon em "Tiers" de poder, com base no `Base Stat Total`.

In [ ]:
# 1. Primeiro, criamos nossa função com as regras de classificação
def classificar_poder(total_de_stats):
    if total_de_stats > 550:
        return 'Nível S (Super Raro)'
    elif total_de_stats > 450:
        return 'Nível A (Forte)'
    elif total_de_stats > 300:
        return 'Nível B (Comum)'
    else:
        return 'Nível C (Básico)'

In [ ]:
# 2. Agora, usamos .apply() para aplicar essa função em cada valor da coluna 'Base Stat Total'
df['Power Tier'] = df['Base Stat Total'].apply(classificar_poder)
df[['Name', 'Base Stat Total', 'Power Tier']].sample(20)

,Name,Base Stat Total,Power Tier
164,mr-mime-galar,460,Nível A (Forte)
385,electrike,295,Nível C (Básico)
997,centiskorch,525,Nível A (Forte)
88,machop,305,Nível B (Comum)
256,slowking,490,Nível A (Forte)
69,meowth-alola,290,Nível C (Básico)
514,floatzel,495,Nível A (Forte)
799,furfrou,472,Nível A (Forte)
467,registeel,580,Nível S (Super Raro)
661,krokorok,351,Nível B (Comum)


### **8. Hora dos Desafios!**

Agora é sua vez de praticar seus novos poderes de Mestre de Dados!

**Desafio 1: A Elite de Kanto** Use um filtro composto (`&`) para criar um novo DataFrame chamado `elite_kanto` que contenha apenas Pokémon da `generation-i` **E** que tenham um `Base Stat Total` maior que 500. Mostre os 5 primeiros desse novo DataFrame.

**Desafio 2: Contagem de Gerações** Qual geração (`Generation`) tem o maior número de Pokémon? Use o `.value_counts()` para descobrir e mostrar a lista.

**Desafio 3: O mais Rápido ou o mais Forte?** Crie um DataFrame chamado `rapidos_ou_fortes` que contenha todos os Pokémon que tenham `Speed` maior que 120 **OU** `Attack` maior que 120. Mostre uma amostra aleatória de 5 Pokémon desse grupo.

**Desafio 4: Classificação de Peso** Usando `.apply()` e uma função que você mesmo vai criar, crie uma nova coluna chamada `Categoria de Peso`. As regras são:
- Se `Weight (hg)` for maior que 2000, classifique como "Super Pesado".   
- Se for entre 500 e 2000, classifique como "Pesado".
- Se for menor que 500, classifique como "Leve". Mostre a nova coluna junto com 'Name' e 'Weight (hg)'.


    

**Desafio 5: Análise dos Psíquicos** Use o `.describe()` para ver um resumo estatístico (`Attack`, `Defense`, `Speed`) apenas dos Pokémon do tipo Psíquico (`Primary Typing == 'psychic'`). Será que eles são, em média, mais rápidos ou mais fortes que a média geral que vimos no início da aula?

Parabéns por chegar até aqui! Você está cada vez mais perto de se tornar um verdadeiro Mestre de Análise de Dados!